### DISCUSSION TITLE
Central Flux instance vs. HelmRelease vs. common base files

### DISCUSSION THREAD

**Discussion by TomyLobo** [fluxcd/flux2#3494] [Q&A] [answered]
I'm using a central Fluxv2 instance on a management cluster and I have 2 additional clusters: app1 and app2.
All 3 clusters have the same infrastructure components, with (for simplicity's sake) the same configuration.

Relevant for this are 3 Flux resources for each cluster:
1. Kustomization A, which sits at the top level for each cluster and rolls out 2 lower-level Kustomizations for apps and infrastructure.
2. Kustomization B, which is the lower-level Kustomization for the infrastructure components.
3. HelmRelease C, which installs one of the infrastructure components, external-dns.

The flux-system Kustomization of the management cluster rolls out A.
A rolls out B.
B rolls out C.

A and B are in cluster-specific files, C is in a common base file.

In order to roll out infrastructure components on the app1/2 clusters, I need to add a kubeConfig key to the spec of one of the 3 resources.
If I add kubeConfig to A, I get:
```
Kustomization/flux-system/app1-infrastructure dry-run failed, error: no matches for kind "Kustomization" in version 
"kustomize.toolkit.fluxcd.io/v1beta2"
```
If I add kubeConfig to B, I get:
```
HelmRelease/component1/component1 dry-run failed, error: no matches for kind "HelmRelease" in version "helm.toolkit.fluxcd.io/v2beta1" Namespace/component1 created
```
If I add kubeConfig to C, I'm overwriting the HelmRelease objects for the other clusters, since they're the same cluster+namespace+name.

The errors in cases 1 and 2 are due to the remote clusters missing Flux's CRDs, so I considered installing those on the remote clusters.
But without a Helm Controller on those clusters, my HelmRelease wouldn't get installed, right?
Is there a way to have all 3 of these:
1. Just one central Flux instance
2. Rolling out a Helm release via Flux on a app1/app2
3. Describing that Helm release with a common base file for management/app1/app2

Maybe Flux isn't designed for this, but it seems like a very basic use case of the multi-cluster feature.
So I hope I'm the problem and not Flux :)

**Comment by Stringls** [comment]
Hi! Yeah, I also faced up with this issue. I tried to use centralized flux on the mgmt cluster, but I am not able to deploy resources from Flux CRDs on workload clusters because they don't have Flux installed. And I am wondering what's the benefit of having centralized Flux if it is not able to deploy HelmRelease, etc. onto workload clusters?

**Reply by kingdonb** [reply]
It can! I think you both must not have seen the `kubeConfig` field!

I mentioned this also in the issue that @TomyLobo opened, but yes, this is absolutely a feature of Flux, you just need to have a `KUBECONFIG` file that has the appropriate permissioned token in it so that you can create the resources you need on the target cluster. This is supported in both `HelmRelease` and `Kustomization`.

**Reply by kingdonb** [reply]
https://fluxcd.io/flux/components/kustomize/kustomization/#remote-clusters--cluster-api
A similar field exists also for HelmReleases. Please check it out.

**Reply by Stringls** [reply]
> https://fluxcd.io/flux/components/kustomize/kustomization/#remote-clusters--cluster-api A similar field exists also for HelmReleases. Please check it out.

Ah, I didn't know it works for `HelmReleases` as well, thanks.
Yeah, it seems good, but I think I won't be able to deploy charts from the private repositories. By default I reference to the private `GitRepository` with this helm chart, but how to deploy this `GitRepo` to workload clusters if this feature is only supported in `HelmRelease` and `Kustomization`?

**Reply by TomyLobo** [reply]
@kingdonb I have seen that field, but that doesn't help, see what I wrote in the original post:
> If I add kubeConfig to [the HelmRelease] C, I'm overwriting the HelmRelease objects for the other clusters, since they're the same cluster+namespace+name.

I have 1 base file that contains the base HelmRelease definition.
As far as I know, there are 2 fields I cannot override: name and namespace
So if I roll out the same file for 2 clusters and I'm using the central approach, both times, the HelmRelease object ends up on the management cluster and since they both have the same name+namespace, there's a conflict

**Reply by kingdonb** [reply]
@Stringls The workload pulls the source from the local cluster. You can deploy HelmReleases with a GitRepository or HelmRepository source, and Kustomizations with any source, from a parent cluster to a target cluster without installing any Flux on the target clusters. Only the workloads will be deployed onto the target cluster, then Kubernetes takes the hand-off and manages a workload from there.

I think there may be more issues, as I have read in the other thread that I did not understand completely the framing of the issue posed by @TomyLobo and there's a separate issue. We're not just trying to install a `HelmRelease` onto a target cluster, we're trying to install the same `HelmRelease` onto many target clusters. Because of the way that reconciling manages a resource at a time, that's not going to work. That might be separate from your issue, @Stringls if I understand what you're saying, I suggest you open a separate thread so we do not conflate the two issues.

**Reply by Stringls** [reply]
@kingdonb Thank you for explnations. I tried this approach and it works perfectly.
I don't think there is a need for a new issue anymore :)

**Comment by TomyLobo** [comment, marked as answer]
I opened an issue from this:
https://github.com/fluxcd/flux2/issues/3499
